### PR TITLE
Add Readme Stream

### PR DESCRIPTION
**Pull request by Ry-DS**
An attempt at adding the README stream.
A few questions
- What do I do for the replication key and primary key?
- Is it ok to just return the json given for the parse_response? We aren't getting a list or anything, so I think it's fine here.
- poetry isn't making a .venv, and is instead storing it somewhere in my user app data. Is this special for this project? 
image
Just a little confused since the oviohub repo adds it as a .venv, and having different behaviour here is confusing me haha

### CODE REVIEW

File: tap_github/streams.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 from typing import Any, Dict, Iterable, List, Optional
 
+import requests
```
Comment (laurentS) [inline tap_github/streams.py:5]: This tells me you forgot to pull the latest `main` branch before writing your code ;)

File: tap_github/streams.py [hunk 2]
```diff
@@ -92,6 +93,42 @@ def get_child_context(self, record: dict, context: Optional[dict]) -> dict:
...
+        th.Property("path", th.StringType),
+        th.Property("content", th.StringType),
+        th.Property("sha", th.StringType),
+        th.Property("url", th.StringType),
+        th.Property("git_url", th.StringType),
+        th.Property("html_url", th.StringType),
+        th.Property("download_url", th.StringType),
+        th.Property(
+            "_links",
+            th.PropertiesList(
```
Comment (laurentS) [inline tap_github/streams.py:123]: I think you need `th.Object(...` here instead of `th.PropertiesList`.

**Review by laurentS — CHANGES_REQUESTED** [review]
Does the code work as is? the schema definition bothers me a bit :)

Probably the best test would be to run this within the ovio cli with the `readme` stream and see if you get the data in the db (if you use the code as is at the time of this comment, you will get a final error about invalid state, that's not related to your changes). If so, I think you're good to go and open a PR to the main meltano repo.
But fetch the upstream changes first ;)

**Comment by laurentS** [comment]
>     * What do I do for the replication key and primary key?

I think the readme stream is like the `repositories` stream, ie: not incremental, but "full table" replication, meaning you fetch the "entire stream" each time you run it (which in this case is a single record). So you don't actually need any `replication_key*` attributes, since these define the incremental behaviour.

>     * Is it ok to just return the json given for the parse_response? We aren't getting a list or anything, so I think it's fine here.

I think you don't even need to implement `parse_response`, unless the api's response is not a list, like for `repositories`. 

>     * poetry isn't making a .venv, and is instead storing it somewhere in my user app data. Is this special for this project?
>       image
>       Just a little confused since the oviohub repo adds it as a .venv, and having different behaviour here is confusing me haha

Fair enough. This is actually poetry's default behaviour, which we modified in the other venvs to be [in-project](https://python-poetry.org/docs/configuration/#virtualenvsin-project) mostly because heroku's deployment system was messing with poetry's logic. If it bothers you while developing (it does mean that your dependencies' code is in this random place), you can copy `poetry.toml` from the main repo over to the tap's repo, remove the venv you already have with `poetry env remove 3.9` (or whatever version you're using) ([docs](https://python-poetry.org/docs/managing-environments/#deleting-the-environments)) and then rerun `poetry install` which will then put the venv under `.venv`. Just don't commit the file since that doesn't seem to be part of meltano's logic.

**Review by Ry-DS — COMMENTED** [review]
> Does the code work as is? the schema definition bothers me a bit :)
> 
> Probably the best test would be to run this within the ovio cli with the `readme` stream and see if you get the data in the db (if you use the code as is at the time of this comment, you will get a final error about invalid state, that's not related to your changes). If so, I think you're good to go and open a PR to the main meltano repo.
> But fetch the upstream changes first ;)

Tried running, and it works!. Can confirm since I get the really long encoded README `content` in the console when I run `tap-github --config config-sample.json`. I'll need to check all fields properly tho. I'll do this on Ovio to make sure :+1:
Sample output:
image

**Comment by Ry-DS** [comment]
> I think you don't even need to implement parse_response, unless the api's response is not a list, like for repositories.

If I didn't I got an error: `TypeError: 'NoneType' object is not iterable`
This is because the default parse_response expects an items property: https://github.com/oviohub/tap-github-meltano/blob/052377d6043f71e2c3ff7045573ebad68d702072/tap_github/client.py#L80
Should I leave my current impl as is?

**Comment by laurentS** [comment]
> > I think you don't even need to implement parse_response, unless the api's response is not a list, like for repositories.
> 
> If I didn't I got an error: `TypeError: 'NoneType' object is not iterable`
> This is because the default parse_response expects an items property: https://github.com/oviohub/tap-github-meltano/blob/052377d6043f71e2c3ff7045573ebad68d702072/tap_github/client.py#L80
> Should I leave my current impl as is?
> 
> Thanks for answering my other questions btw :)
> 
> 

I think you can do what I did for repos, just `return [row]` or whatever the variable is. That way you'll have an iterable.

File: tap_github/streams.py (hunk 1 above)
Comment (laurentS) [reply tap_github/streams.py:6]: It's just there 3 lines above, no need to duplicate it :)

**Comment by Ry-DS** [comment]
Opened https://github.com/MeltanoLabs/tap-github/pull/15

File: tap_github/streams.py (hunk 1 above)
Comment (Ry-DS) [reply tap_github/streams.py:3]: ahh missed that for some reason haha. Thanks for the catch

**Comment by ericboucher** [comment]
Merged and pulled from parent repo.

**Pull request closed without merging**